ksql-rest-app/src/test/java/io/confluent/ksql/rest/integration/HeartbeatAgentFunctionalTest.java

```diff
@@ -0,0 +1,265 @@
+    final ClusterStatusResponse clusterStatusResponseUp = waitForRemoteServerToChangeStatus(
+        this::remoteServerIsUp);
+
+    // Then:
+    assertThat(clusterStatusResponseUp.getClusterStatus().get(host0.toString()).getHostAlive(), is(true));
+    assertThat(clusterStatusResponseUp.getClusterStatus().get(host1.toString()).getHostAlive(), is(true));
+  }
+
+  @Test(timeout = 60000)
+  public void shouldMarkRemoteServerAsDown() {
+    // Given:
+    waitForClusterToBeDiscovered();
+
+    // When:
+    ClusterStatusResponse clusterStatusResponse = waitForRemoteServerToChangeStatus(
+        this::remoteServerIsDown);
+
+    // Then:
+    assertThat(clusterStatusResponse.getClusterStatus().get(host0.toString()).getHostAlive(), is(true));
+    assertThat(clusterStatusResponse.getClusterStatus().get(host1.toString()).getHostAlive(), is(false));
+  }
+
+  @Test(timeout = 60000)
+  public void shouldMarkRemoteServerAsUpThenDownThenUp() {
+    // Given:
+    waitForClusterToBeDiscovered();
+    sendHeartbeartsEveryIntervalForWindowLength(100, 2000);
+
+    // When:
+    final ClusterStatusResponse clusterStatusResponseUp1 = waitForRemoteServerToChangeStatus(
+        this::remoteServerIsUp);
+
+    // Then:
+    assertThat(clusterStatusResponseUp1.getClusterStatus().get(host0.toString()).getHostAlive(), is(true));
+    assertThat(clusterStatusResponseUp1.getClusterStatus().get(host1.toString()).getHostAlive(), is(true));
+
+    // When:
+    ClusterStatusResponse clusterStatusResponseDown = waitForRemoteServerToChangeStatus(
+        this::remoteServerIsDown);
+
+    // Then:
+    assertThat(clusterStatusResponseDown.getClusterStatus().get(host0.toString()).getHostAlive(), is(true));
+    assertThat(clusterStatusResponseDown.getClusterStatus().get(host1.toString()).getHostAlive(), is(false));
+
+    // When :
+    sendHeartbeartsEveryIntervalForWindowLength(100, 2000);
+    ClusterStatusResponse clusterStatusResponseUp2 = waitForRemoteServerToChangeStatus(
+        this::remoteServerIsUp);
+
+    // Then:
+    assertThat(clusterStatusResponseUp2.getClusterStatus().get(host0.toString()).getHostAlive(), is(true));
+    assertThat(clusterStatusResponseUp2.getClusterStatus().get(host1.toString()).getHostAlive(), is(true));
+  }
+
+  private void waitForClusterToBeDiscovered() {
+    while (true) {
+      final ClusterStatusResponse clusterStatusResponse = sendClusterStatusRequest(REST_APP_0);
+      if(allServersDiscovered(clusterStatusResponse.getClusterStatus())) {
+        break;
+      }
+      try {
+        Thread.sleep(200);
+      } catch (final Exception e) {
+        // Meh
+      }
+    }
+  }
+
+  private boolean allServersDiscovered(Map<String, HostStatusEntity> clusterStatus) {
+    if(clusterStatus.size() < 2) {
+      return false;
+    }
+    return true;
+  }
+
+  private void sendHeartbeartsEveryIntervalForWindowLength(long interval, long window) {
+    long start = System.currentTimeMillis();
+    while (true) {
```

Comment:
```suggestion
    while (System.currentTimeMillis() - start < window) {
```
and remove the break condition